compiler/hash-typecheck/src/diagnostics/error.rs

```diff
@@ -16,14 +19,24 @@ pub type TcResult<T> = Result<T, TcError>;
 pub enum TcError {
     /// Cannot unify the two terms.
     CannotUnify { src: TermId, target: TermId },
+    /// Cannot unify the two argument lists. This can occur if the names
+    /// don't match of the arguments or if the number of arguments isn't the
+    /// same.
+    CannotUnifyArgs {
+        src_args_id: ArgsId,
+        target_args_id: ArgsId,
+        src: TermId,
+        target: TermId,
+        reason: ParamUnificationErrorReason,
+    },
     /// Cannot unify the two parameter lists. This can occur if the names
     /// don't match of the parameters or if the number of parameters isn't the
-    /// same.
+    /// same, or the types mismatch.
     CannotUnifyParams {
         src_params_id: ParamsId,
         target_params_id: ParamsId,
-        src: TermId,
-        target: TermId,
+        src: LocationTarget,
+        target: LocationTarget,
         reason: ParamUnificationErrorReason,
     },
```

Comment:
these could probably be one variant

Reply:
You mean src and target location targets?

Reply:
No, I mean both variants `CannotUnifyArgs` and `CannotUnifyParams`, this was the idea behind `ParamListKind` to unify these kinds of errors (more specifically the error reporting logic)

Reply:
Ah yeah that would be nice.. I'll see what I can do

Reply:
It's actually not too easy to do this.. I think it is better left to another PR.

Reply:
See #415